### DISCUSSION TITLE
Genetic optimization design

### DISCUSSION THREAD

**Discussion by cuda** [mathnet/mathnet-numerics#792] [Ideas]
I've added a simple genetic optimization solver for continuous functions to my fork -
https://github.com/cuda/mathnet-numerics/tree/ga

I'm not sure of the design. I used delegates for the primary functions (pairing, mating, mutating, evaluating/cost, and convergence) so users could pass in lambdas.

Simple example using defaults (minimize x^2 where  -10 < x < 10):

```
var solver = new ContinuousSolver((double[] x) => x[0]*x[0], ContinuousBounds.CreateBounds(1, -10, 10));
var solutions = solver.Evolve();
```

The solver takes in a cost function as a delegate and an IList of bounds.

Users can override the default behavior by setting the various delegates using lambdas (or they could use regular methods):

```
solver.Pair = (options, population) => ...
solver.Crossover = (options, bounds, ma, pa) => ...
solver.Mutate = (options, bounds, population) => ...
solver.Converge = (options, currentStatus, population) => ...
```

I prefixed the solver and bounds with Continuous so we could support other function types in the future, but since they are already in the Continuous namespace it might be redundant. 

Any comments on the design or code would be greatly appreciated.

Thanks,
Marcus

**Comment by cdrnet** [comment]
Looks good, thanks for bringing it over! I actually have a use case for this right now, I'll get back with some feedback on my experience on using it in practice.
